[Core] Implements prepareStyles as composition of functions in muiTheme

newoga opened the pull request:
This commit implements the `prepareStyles` function as a composition of functions that can be configured in `muiTheme`. It switches the `./mixin/style-propable.js` implementation to use the `prepareStyles` stored in context.

@oliviertassinari @alitaheri 
This is an initial implementation of the approach of storing `prepareStyles` in context. I cleaned up the code a bit, however like I said, I don't think we should document this until we've confirmed we're happy and the API is final (I already have some ideas of how to improve it, but it shouldn't have any real immediate benefit). It's also very well possible this will all go away when we adopt a new styling framework.

**Note**: One minor difference I made is that the `getMuiTheme` function actually checks the computed theme for `isRtl` before composing the functions. If it's false, it doesn't even bother including the `ensureDirections()` (renamed to `rtl()`) method. **Edit:** It now does this for the `autoprefixer()` method based on the `userAgent` property

~~Right now, the list of transformers can be overridden by adding the following property to the muiTheme.  The following configuration would essentially disable the default behavior of always autoprefixing and calling `rtl()` when `isRtl` is equal to `true`.~~ Removed customizing of transformers in this PR, other approaches have since been implemented.

**Remaining tasks:**
- [x] Move implementation of "double call warning" out of `rtl()` into something that runs before all style transformations are executed. **Done**
- [x] ~~Change `rtl()` implementation to not care is `isRtl`~~ **No longer necessary**

**Future PR:**
- Convert components not using `./mixins/style-propable.js` to get `prepareStyles()` from context, not from importing it from `./utils/style.js`.
- Deprecate and/or remove `./utils/style.js`

newoga commented:
Related to #2316, #2530, #2852.

alitaheri commented:
@newoga Awesome work indeed :+1: :+1: Thank you :smile:

alitaheri commented on src/compose.js:
```diff
@@ -0,0 +1,3 @@
+import flowRight from 'lodash.flowright';
```
Do we need these 2 files? I mean it's gonna look like "API surface" to me. We can just require these whenever we want. and I'm not seeing a day we wouldn't use lodash, you know, since it's awesome :grin:

newoga replied:
@alitaheri I had the same though and I'm open to removing these. The only library I thought we may one day replace these with is [recompose](https://github.com/acdlite/recompose), which ironically also uses lodash's compose for its implementation. 

Either way I'm going to take it out `compose.js` until we start using it in several places. 

I wanted to talk to you two about `merge.js` though. I'm thinking we should replace all `mergeStyles` with one of the three:

1. merge
2. Object.assign
3. destructuring

alitaheri replied:
how about `lodash.assign`? It uses Object.assign if it's available. and it's faster than extend or merge imo. @oliviertassinari Thoughts?

newoga replied:
For the record I'd be happy with that :+1:

alitaheri commented on src/styles/getMuiTheme.js:
```diff
@@ -224,4 +245,11 @@ export default function getMuiTheme(baseTheme, muiTheme) {
+
+  return merge(computedTheme, {
+    prepareStyles: createPrepareStyles(computedTheme, transformers),
+  });
```
why not just:

```js
computedTheme.prepareStyles = createPrepareStyles(computedTheme, transformers);
return computedTheme;
```
?

newoga replied:
Because I'm stupid :laughing: 

Thanks :+1:

alitaheri replied:
I've done worse myself :sweat_smile: :sweat_smile:

alitaheri commented on src/styles/getMuiTheme.js:
```diff
@@ -224,4 +245,11 @@ export default function getMuiTheme(baseTheme, muiTheme) {
   }, muiTheme);
+
+  const transformers = computedTheme.transformers ? computedTheme.transformers :
+    computedTheme.isRtl ? [autoprefixer, rtl] : [autoprefixer];
```
please break this down to if else chain. It's hard to read :fearful:

newoga replied:
Done :smile:

alitaheri commented on src/styles/getMuiTheme.js:
```diff
@@ -224,4 +245,11 @@ export default function getMuiTheme(baseTheme, muiTheme) {
     },
   }, muiTheme);
+
+  const transformers = computedTheme.transformers ? computedTheme.transformers :
```
I feel like it would be more readable to get the transformers as function argument. 

like:

```js
getMuiTheme(baseTheme, muiTheme, transformers)
```
Since people are very unlikely to want to modify transformers.

newoga replied:
I agree, I'm a little scared to make it this exposed yet :fearful:

alitaheri replied:
it's `js` adding arguments isn't exposing :laughing: :laughing:

newoga replied:
Haha! Putting the transformers as a function argument, as you suggested, was also my first thought, and I was trying really really hard not to do it :laughing: 

I suppose part of me is really wanting to make sure this feature isn't abused and isn't going to be something 3 months from now we're trying to figure out how to deprecate properly. I know I'm probably being over paranoid, but it makes me feel better if someone and everyone else agrees that it's okay. I'd rather assume what I'm doing is unsafe.

alitaheri replied:
it works either way. I'm alright with it. With something like C# I would be paranoid too since IL has `hide by signature`, in other words signature is documentation. but with a language this dynamic. I don't see any difference actually i just think this is easier:

```js
getMuiTheme(null, null, [stuff, stuff2]);
```

than

```js
getMuiTheme(null, {transformers: [stuff, stuff2]});
```

newoga replied:
Yeah, I generally agree with you. Though I suppose it's easier to change the implementation of where transformers are applied as a non breaking change when they are in an object rather than set arguments that rely on order/positioning. 

Also, since right now all of the existing arguments are being merged together, the transformers kind of break pattern a bit. I am going to put it in the arguments since you agree that it is better and we can see how @oliviertassinari feels.

alitaheri replied:
> the transformers kind of break pattern a bit

Yeah, you have a point there :grin:

alitaheri commented on src/styles/transformers/rtl.js:
```diff
@@ -0,0 +1,90 @@
+ * checking `styleConstants` in `muiTheme` and replacing attribute keys if
+ * necessary.
+ */
+export default function rtl(muiTheme, style) {
```
how about: 

```js
export default theme => style => {
//...
};
```

This will be like a transformer factory function and it will dramatically simplify your `createPrepareStyles` function, which took me some time to understand :grin:. since you won't need to wrap before chain complication the behavior.

newoga replied:
Yes I originally had that in one of my earlier designs, I forget why that was an issue. Mmm...

newoga replied:
Oh right, I didn't like the idea that theme could be modified (and changes passed) through the transform chain. Though I saw that redux middleware does something to the sort this (where getTheme is getState):

```js
export default ({getTheme}) => style => {
// getTheme is a function that returns the theme
};

alitaheri replied:
Why could it be modified? We are treating it as immutable. so we shouldn't worry about that right?

newoga replied:
Mm, actually for some reason I'm having trouble seeing how this could work: 

```js
export default theme => style => {
//...
};
```
If I'm not mistaken, since the `compose` function is taking a single return value and passing it as an argument to the next function in the chain, if we composed a set of functions with this signature, then wouldn't the second function in the chain receive the style in the theme?

I think the only way we can properly use compose with multiple objects such as theme and style, is to package them in an object together and make that the combined object the return and input values for the function.  Does that make sense?

alitaheri replied:
or

```js
function createPrepareStyles(muiTheme, transformers) {
  return compose(...transformers.map(t => t(muiTheme));
};
```

newoga replied:
@alitaheri :astonished:

newoga replied:
What the heck! I swear I had something like this and it wasn't working!!!!! Thank you!

This can work as beautifully as I hoped.

alitaheri replied:
I hope it does work... Please work. don't embarrass me, idea :pray: :pray:

alitaheri commented on src/styles/getMuiTheme.js:
```diff
@@ -224,4 +245,14 @@ export default function getMuiTheme(baseTheme, muiTheme) {
     },
   }, muiTheme);
+
+  if (computedTheme.transformers) {
```
Beautiful :heart_eyes: :heart_eyes:

newoga replied:
I like it too :+1: Thanks! :dancers:

newoga commented on src/styles/getMuiTheme.js:
```diff
@@ -224,4 +226,12 @@ export default function getMuiTheme(baseTheme, muiTheme) {
     },
   }, muiTheme);
+
+  if (computedTheme.isRtl) {
```
@alitaheri How do you feel about this?

The only reason why this is slightly different than before is that the `rtl` transformer will be added to the list of transforms is `isRtl` is set to true (regardless of whether or not the specified their own custom list of transformers). I think I'm okay with this because `isRtl` is false by default.

Also, I'm thinking that if we make this a documented API we should eventually deprecate `isRtl` in favor of them just applying the transformer themselves.

alitaheri replied:
> hat the rtl transformer will be added to the list of transformers if isRtl is set to true

This is a bit scary. :P What if they want to add transformation after rtl is applied?

> Also, I'm thinking that if we make this a documented API we should eventually deprecate isRtl in favor of them just applying the transformer themselves.

I think some components rely on `isRtl`.

newoga replied:
> I think some components rely on isRtl.

Yikes :scream_cat: , that point I did not realize. I thought the whole point of `isRtl` was so that components didn't have to be written in a way where they worried rtl vs ltr, and that it would be properly converted in prepareStyles.

Mm, I think we should consider removing the dependencies on `isRtl` from the components.

> This is a bit scary. :P What if they want to add transformation after rtl is applied?

In my view, if they wanted to do this, they would just simply supply their own list of transformers `[autoprefix, customTransformer, rtl]` and just set `isRtl` to false. In other words, `isRtl` is only for backwards compatibility purposes.

Now because of your other point regarding components dependent on `isRtl` that couldn't work.

newoga replied:
> after rtl is applied?

Actually technically the issue would be if they wanted transformers to run *before* rtl is applied. If we push at the end of the list as the current implementation is doing it, it will be the first one ran, then custom transformers (since we're using lodash's `flowRight` instead of `flow`)

alitaheri replied:
Right right :grin: I'm really bad with left and right, before and after :D 

> I think we should consider removing the dependencies on isRtl from the components.

I think this will be very limiting. Some logics related to rtl is simply not implementable with style modification. I wouldn't change this behavior :grin:

oliviertassinari replied:
> I think this will be very limiting. Some logics related to rtl is simply not implementable with style modification. 

I completely agree, `isRtl` is affecting the style but also the component order and which component we use.
I think it's fine like this.

oliviertassinari commented on src/merge.js:
```diff
@@ -0,0 +1,3 @@
+import merge from 'lodash.merge';
+
+export default merge;
```
What about using `Object.assign` with `babel-plugin-transform-object-assign`? (a96dcb8a62499f3039dee03a6a548841d96065e3)

alitaheri replied:
We did agree on this on an diffed out discussion. lodash.assign seems like a better choice.

newoga replied:
Mm, I wouldn't mind a babel plugin that injects `Object.assign` either. I guess the benefit of `Object.assign` is it's standard ES6 syntax so we won't have to worry about imports.

@alitaheri @oliviertassinari I agree with `assign` over `merge`, but do either of you feel strongly one way or the other lodash v object.assign w/ babel?

alitaheri replied:
I don't know about babel. but I know lodash will use Object.assign if available, if not it's own implementation. native Object.assign might be faster? I don't know, nor do i think it can actually have any impact what so ever. I use lodash everywhere, and it has never let me down. But babel doesn't seem bad either. I'm alright with both approaches :P

oliviertassinari replied:
Sorry, I haven't catched up the conversation. So, you said
> how about lodash.assign? It uses Object.assign if it's available. and it's faster than extend or merge imo. @oliviertassinari Thoughts?

The solution I'm proposing is using the babel `_extend`. This babel function is using `Object.assign` when available or a polyfill. In my case, It was reducing the output size of the lib. The `_extend` was already needed for the spread operator syntax. In our case, `lodash.merge` is probably better :+1:.

alitaheri replied:
> In our case, lodash.merge is probably better

Do you mean `lodash.assign`? O.o

oliviertassinari replied:
I'm talking about `import merge from 'lodash.merge';`. I guess I shouldn't comment in my state :beer: :grin:.

alitaheri replied:
well merge is a bit more expensive, assign does explicit shallow merge. so I guess it's better.

But for now... Go drunk, you're home!

alitaheri commented:
I believe we should use assign for every style merging too. :+1: Thanks @newoga

oliviertassinari commented on src/styles/transformers/rtl.js:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  // Left to right is the default. No need to flip anything.
+  if (!muiTheme || !muiTheme.isRtl) return style;
```
I think that we don't need this line.
More generally, do we really need to provide the `muiTheme` to the transformers?

alitaheri replied:
@oliviertassinari Let's not limit ourselves for now. We might want to add other things, or if this is gonna be documented, others might like to have access to the theme in their transformers. the impacts of having it is likely none. but excluding it completely might make it hard to put back (or maybe even impossible). besides! we might even be able to use this to pass down userAgent down the context and use it in the prefixer. Just saying :grin:

newoga replied:
@oliviertassinari I agree we don't need the line but also agree with @alitaheri  that we should possibly leave the muiTheme in the transformers for now.

The reason why the `muiTheme` was there in the first place is because this `ensureDirections` now names (rtl) was left untouched minus the name and merge implementation. I believe the rtl check could be removed and the warning show be moved top level too (these are currently listed as remaining task at the top of the PR)

oliviertassinari commented on src/styles/transformers/rtl.js:
```diff
@@ -0,0 +1,90 @@
+ * necessary.
+ */
+export default (muiTheme) => (style) => {
+  if (process.env.NODE_ENV !== 'production') {
```
What about creating a third transformer that would not be added in production to makes sure we call `prepareStyles` only once?

alitaheri replied:
But then, if others wish to customize the transformers they should know about that extra transformer and also know that that other transformer shouldn't go into production builds.

oliviertassinari replied:
> also know that the other transformer shouldn't go into production builds

It would be something private that people wouldn't be aware of.

alitaheri replied:
I don't think it would be possible to hide it. Unless we do some weird logic around the transformers array to figure out if it has rtl and push it before it into the array. I can't think of anything else O.o

oliviertassinari replied:
I was thinking about doing something like:
```js
if (process.env.NODE_ENV !== 'production') {
  transformers.push(prepareCallOnlyOnce);
}
```

alitaheri replied:
hmmm so you mean a transformer that check for every transformer instead of only for rtl... :+1: :+1: Absolutely agree. Love that idea :+1: :+1:

newoga replied:
@oliviertassinari @alitaheri Yup, I like that plan too.

newoga commented:
@oliviertassinari @alitaheri 

I made some additional changes to this and rebased onto master so that it includes @oliviertassinari's auto-prefix changes. I decided to take passing in custom transformers out of the API for now. I think having that feature is now less urgent because the the two existing transformers (prefixing and RTL) can be disabled through `muiTheme` configuration (`userAgent` set to false and `isRtl` set to false disables both).

If we want to explore making this publicly accessible, I think we can explore options in a new PR.

The two key changes I made were the following:

##### Moved the check for running `prepareStyle()` twice out of the `rtl()` function

I decided to not to make the check a transformer that was part of the chain, because there would be no way to cancel the rest of the chain from running. I decided instead to wrap the chained function with the check so that it gets called first and exits if it sees if it was ran already.

In addition, I changed the check from using a object key to a `Symbol`. I think this is safer so that developers can't set the property to "trick" the check into running and it also gets the nasty `did-flip` keys out of the style objects in the DOM.

The end result is that this check is ran in development AND production, just that it only logs a warning in development. I think this is the better approach though.

##### Changed `mergeStyles()` implementation to `Object.assign()`.

The first reason is because I think it's simpler, but also `Object.assign()` supports assigning `Symbol` keys and out previous merge implementation did not. This change was required for the first step above. I think we were all leaning towards moving towards either `Object.assign()` or lodash's `assign` anyway.

alitaheri commented:
@newoga I see you have brainstormed the crap out of it :+1: :+1: Really great job. I really like your approaches :+1: :+1: :tada:

alitaheri commented on src/utils/styles.js:
```diff
@@ -1,16 +1,11 @@
-  if (!objB) return objA;
-  return update(objA, {$merge: objB});
-}
+export const mergeStyles = (...args) => Object.assign({}, ...args);
```
Do we really need to clone? I think it's an unnecessary allocation of an extra object. @oliviertassinari Thoughts?

newoga replied:
I wish we didn't need this but it was breaking things when we weren't cloning. I think we have some situations in source where we may be passing in an undefined style, but I haven't been able to pinpoint everywhere it's happening yet. I just know calling simply `Object.assign` broke the docs site and the tests.

I think it's worth going back and investigating though! I think the ideal would be that we shouldn't need to do this.

newoga replied:
Oh and one more point. the `mergeStyles` is really going to be deprecated/removed when `style-propable` is removed. This is kind of temporary.

I'm recommending we just reimplement the components to use the `prepareStyles` from context directly, not from the `style-propable` mixin, and anywhere we were calling `mergeStyles()` we replace with `Object.assign`. After that this can go away.

alitaheri replied:
Well, not having it is dangerous. Let it be, With memoization and proper optimization this doesn't matter at. So we better keep it like this.

newoga replied:
I agree, sounds good :+1:

alitaheri replied:
> I'm recommending we just reimplement the components to use the prepareStyles from context directly, not from the style-propable mixin, and anywhere we were calling mergeStyles() we replace with Object.assign. After that this can go away.

Oh, right. So all the more reason to ignore what I said :D :D This is great as it is :grin:

oliviertassinari replied:
> Do we really need to clone?

What do you mean by clone? I think that it's totally fine :+1:

alitaheri commented on src/styles/getMuiTheme.js:
```diff
@@ -4,6 +4,21 @@ import ColorManipulator from '../utils/color-manipulator';
+
+const CALLED_ONCE = Symbol('prepareStyles/calledOnce');
+
+const callOnce = (transformer) => (style) => {
```
This is my new favorite function across the library :heart_eyes: :heart_eyes:

newoga replied:
Haha, thank you very much! :bowtie:

oliviertassinari replied:
Can we avoid to call this function in production with `if (process.env.NODE_ENV !== 'production') {`?

newoga replied:
We could, but after thinking about it, I wondered if this is a good idea?

What would happen would be in development it would just return the style, and in production it would continue flipping/autoprefixing with no warning. I think having this check in production makes it more consistent in terms of what you see in the development build.

What do you think?

oliviertassinari replied:
My motivation is simple. This is reducing the performances. That's something we can accept in development but not in production :grin:.

Regarding your points I agree, but are far less important IMHO.

newoga replied:
Mm, I hear you! You're right.

My initial thought was that the performance cost of just returning the object you passed in is *less* than re-running the whole chain again on the same object and would make this approach more performant in production. But that is only true when compare when the function is used twice. If it is just ran once, the check is unnecessary and would reduce performance.

Also, another point I suppose is this function really is primarily for us not for users of the library. So as long as we're catching the warning then in production it should be fine. It'd be interesting if we could just write a test for this and be done with it.

oliviertassinari replied:
> this function really is primarily for us not for users of the library

Exactly :+1:

newoga commented:
> Also, Does the polyfill also assign Symbols?

That's an intersting question. I *assumed* so but assuming can be dangerous :stuck_out_tongue:. I'll have to investigate this more...

alitaheri commented:
This is the polyfill. unfortunately it doesn't support Symbols. but I think this wouldn't matter because if `Object.assign` is not supported then so isn't Symbol. 

```js
var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];
    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }
  return target;
};
```

oliviertassinari commented on src/styles/getMuiTheme.js:
```diff
@@ -4,6 +4,21 @@ import ColorManipulator from '../utils/color-manipulator';
+import compose from 'lodash.flowright';
+import warning from 'warning';
+
+const CALLED_ONCE = Symbol('prepareStyles/calledOnce');
```
Introducing a `Symbol` will have some implication, have a look at the issues that this PR solve https://github.com/callemall/material-ui/pull/2904.
I think that it's a breaking change.

newoga replied:
Mm, interesting. What do you and @alitaheri  prefer? Use a regular string key or polyfill?

I wonder we're inevitably going to keep running into `Symbol` issues because developers often assume it's "included".

alitaheri replied:
I think we should avoid polyfills for the time being. es6 has to become mainstream before we can use it libraries. I use every feature i can get my hands on in my application. but libraries are different :grin:

newoga replied:
Okay `did-flip` from before made more sense in the context of ensureDirection. Do you have a preference to what key we use? `prepared: true`?

oliviertassinari replied:
What about `mui-prepared: true`?
We should consider using a `Symbol` again with v0.15.0.

oliviertassinari commented on src/styles/transformers/rtl.js:
```diff
@@ -0,0 +1,78 @@
+ */
+export default (muiTheme) => (style) => {
+  // Left to right is the default. No need to flip anything.
+  if (!muiTheme || !muiTheme.isRtl) return style;
```
Can't we remove this check?

newoga replied:
Yeah, technically! I left it like this for now for consistency with auto-prefix because the auto-prefix transformer is also technically checking `muiTheme` for on/off behavior. I have an idea on how we can get the best of both worlds:

Something like this:
```js
export default function(muiTheme) {
  if (!muiTheme || !muiTheme.isRtl) return null;

  return (style) => {
    // impl here...
  }
}
```
I have to test, but the compose function might ignore nulls. So that way the transformer can decide based on muiTheme whether or not it should be included in the transfomer list.

alitaheri replied:
Interesting... I like that approach :+1: :+1:

oliviertassinari replied:
> the auto-prefix transformer is also technically checking muiTheme for on/off behavior

I was going to say that it's true but different. We do it only once in the auto prefixer.
I like your proposal.

newoga commented:
@oliviertassinari @alitaheri Alright guys, I think we're close :persevere:

This has the following changes:
1.  Moved all the "transformer inclusion" checks into the transformer definitions themselves. They will be filtered out if they don't return a truthy value.
2. Since we decided that there are performance implications with the `callOnce`, I moved it out to a transformer like @oliviertassinari originally suggested. It returns early if it's in `production` environment.
3. Replaced the `Symbol` with just a string for the callOnce check.

oliviertassinari commented:
@newoga I think we have it! I like the current approach :+1:. We will have to refactor a lot of code once it's merged.

newoga commented:
@oliviertassinari I agree, I think we've made a lot of progress with this. Thanks for your feedback!

Looking forward to getting this merged and working towards getting `style-propable` out. That's definitely the most heavily used mixin at the moment.

alitaheri commented on src/styles/getMuiTheme.js:
```diff
@@ -230,7 +232,9 @@ export default function getMuiTheme(baseTheme, muiTheme) {
     },
   }, muiTheme);
 
+  const transformers = [autoprefixer, rtl, callOnce].filter(t => t(muiTheme));
```
you're initializing the transformers twice. it can have unintended side-effects.

```js
[autoprefixer, rtl, callOnce].map(t => t(muiTheme)).filter(t => t);
// and then
compose(...transformers);
```

Would be safer

oliviertassinari replied:
That makes sense :+1:

newoga replied:
@alitaheri That's much better! Thank you! :+1:

alitaheri commented:
All green here :+1: :green_heart: 

Thanks a lot @newoga This was indeed a work of art :+1: :+1:

newoga commented:
This one was definitely a group effort. Thanks a lot for the help guys :+1:

oliviertassinari commented:
Any idea why this `Popover` example is broken?
![capture d ecran 2016-01-26 a 11 19 54](https://cloud.githubusercontent.com/assets/3165635/12578063/d583e0e2-c41e-11e5-97d9-c8996ff65c98.png)

newoga commented:
> Any idea why the RaisedButton examples are broken?

Wow. Really really good find. This is a really obscure issue but something to consider really carefully. The issue is here:

https://github.com/callemall/material-ui/blob/master/src/enhanced-button.jsx#L285

If that line was changes to:

```js
appearance: linkButton ? null : 'button',
```

Then the issue goes away. 

It has to do with difference in handling nulls and undefined with `Object.assign` vs `merge`. Merge ignores null and undefined and keeps target values and `Object.assign` overwrites.

newoga commented:
I suppose we'll have to decide how much ignoring nulls/undefined matters to us then pick our merge implementation carefully. Ironically, I wonder if this is related to the autoPrefixer issue we were also having for the opposite reason where the animations stopped working with the `all()` method.

oliviertassinari commented:
Setting `appearance: linkButton ? '' : 'button', // edited` doesn't solve the issue on my side. What do you think about my other finding?

newoga commented:
> Setting appearance: linkButton ? '' : 'button', // edited doesn't solve the issue on my side. 

Bah you're right, a linting error messed my build and tricked me. Though I did confirm this is related to `Object.assign`. I thought this was working because I had switched our `mergeStyles` implementation to lodash's merge, and that seemed to fix it for me.

> What do you think about my other finding regarding the hash order?

That's certainly interesting. We can't rule it out yet but I am curious why hash order would make a difference with the styling. I thought order shouldn't matter?

oliviertassinari commented:
> I thought order shouldn't matter?

The order does matter. This is broken:
![capture d ecran 2016-01-26 a 15 32 10](https://cloud.githubusercontent.com/assets/3165635/12583246/001cc26a-c442-11e5-9bbf-361af618ee19.png)
This works:
![capture d ecran 2016-01-26 a 15 33 10](https://cloud.githubusercontent.com/assets/3165635/12583268/212ae8e2-c442-11e5-9cf8-1c157c9ebfcf.png)

newoga commented:
@oliviertassinari Oh oh, I see what you mean now, the keys are different...I misunderstood. Since I thought the keys were the same I figured the previous style value would just get replaced.

newoga commented:
> That's really tricky and linked to how v8 is ordering a hash map.

Yeah, wow. I can't think of a way around this. I'm surprised I haven't seen anyone/any other projects talk about this...

newoga commented:
https://bugs.chromium.org/p/v8/issues/detail?can=1&start=0&num=100&q=getOwnPropertyNames&colspec=ID%20Type%20Status%20Priority%20Owner%20Summary%20HW%20OS%20Area%20Stars&groupby=&sort=&id=4118

oliviertassinari commented:
Well, I don't know exactly why the order change. I'm really curious to understand why. For now. I would say we can this simple fix:
```diff
- background: 'none',
+ background-color: 'transparent',
```

newoga commented:
@oliviertassinari I just tried Firefox and it seems like the issue is actually with Chrome's implementation of `getOwnPropertyNames`. I originally read about this here: https://github.com/sindresorhus/object-assign/issues/22

newoga commented:
>  For now. I would say we can this simple fix:
- background: 'none',
+ background-color: 'transparent',

Yeah that's a good idea.

newoga commented:
@oliviertassinari I was curious and downloaded [Chrome Canary](https://www.google.com/chrome/browser/canary.html) and checked out the commit before I made the style changed and confirmed that the ordering (and buttons) are working properly.

Looks like this was fixed in V8 version [4.9.148](https://chromium.googlesource.com/v8/v8/+log/refs/heads/4.9.148), commit [7d1263d](https://chromium.googlesource.com/v8/v8/+/7d1263db477c812d40789c75be2f368e4c0b9769). If I'm reading [this](https://omahaproxy.appspot.com/) correctly, version `4.9.148` of v8 is currently available in the Chrome development channel. Since there was just a new release of Chrome yesterday, I expect the fix for this should be out in 12 weeks if I'm [understanding their release cycle](https://github.com/v8/v8/wiki/Release%20Process#beta-releases)

oliviertassinari commented:
> Looks like this was fixed in V8 version 4.9.148

That's good to know :+1:. In the mean time. I have found another issue. The `FlatButton` is too small.
![capture d ecran 2016-01-27 a 11 09 26](https://cloud.githubusercontent.com/assets/3165635/12610205/b54dd990-c4e6-11e5-8fbd-e0e54c62eded.png)
I'm wondering if we shouldn't use this ponyfony https://github.com/sindresorhus/object-assign.

alitaheri commented:
I agree with @oliviertassinari object-assign does offer the best implementation.

newoga commented:
@oliviertassinari @alitaheri It looks like that ponyfill just returns `Object.assign` if it already exists so it may not fix the issue (because `Object.assign` does exist in Chrome, it's just not working properly ).  Only other idea is to fork/copy the implementation separately from the `Object.assign || ...` check

alitaheri commented:
@newoga Oh, right there's that :grin: I would think that that's our best bet at the moment.

newoga commented:
> @newoga Oh, right there's that :grin: I would think that that's our best bet at the moment.

@alitaheri @oliviertassinari Would you two be against one of the following approaches:

1. Carefully inspect the rest of the components for times where key ordering is causing a styling issue and investigate approaches like the one above to fixing it (seems like it's only causing issues in isolated cases).
2. Create a babel plugin that does the same thing as `babel-plugin-transform-object-assign` but doesn't use existing native implementation for our project to get around this? (We would remove the plugin after Chromium releases their fixes)

I've given it some though and I've really come around and would prefer to just using `Object.assign` for all the style merging.

alitaheri commented:
Wow, great job investigating this :+1: I think that's a better approach too. Object.assign is more expressive, less cluttered (no need to import) and more es-ish :grin: So I'm all in

oliviertassinari commented:
Both solutions are valid​. Option 2 is definitely quicker to do than option 1. I can't see real advantages of doing option 1 in the long run. I would vote for option 2 :+1:.

newoga commented:
Thanks for the feedback guys. Time to learn babel plugins...:neckbeard:

oliviertassinari commented:
Well, option 3 is to wait a bit before changing the merge function.

newoga commented:
> Well, option 3 is to wait a bit before changing the merge function.

Yup, I considered this too. Though after this gets merged, I wanted to start removing `style-propable`. If the existing `mergeStyles` functions could be replaced with `Object.assign`, we get to drop an import. I'd definitely vote for this if it meant only replacing it in one place such as `utils/style`'s `merge()`but soon all the components would have `Object.assign`

oliviertassinari commented:
> If the existing mergeStyles functions could be replaced with Object.assign

I like this idea. However, using a custom version of `babel-plugin-transform-object-assign` would clearly increase the lib size.

newoga commented:
> I like this idea. However, using a custom version of babel-plugin-transform-object-assign would clearly increase the lib size.

What if it was written in a way that in which the function was imported rather than duplicated each time was used.  If we did that it shouldn't be any bigger than it is now (maybe even smaller). Would you be okay with that?

I haven't completely wrapped my mind around babel plugins but it seems like that it's possible with the little I've looked at it.

oliviertassinari commented:
> What if it was written in a way that in which the function was imported rather than duplicated each time was used.

That would do it :smile:.

newoga commented:
@oliviertassinari @alitaheri 

Take a look at this when you get a chance. Using this approach we're able to fix the issue with the raised-button since we now replace `Object.assign` with an implementation that doesn't using the native implementation.

newoga commented:
@oliviertassinari @alitaheri 

FYI just rebased onto master so this didn't get too far behind. I also decided to remove implementation of mergeStyles to Object.assign in this PR.  The feature presented in this PR (composing style transformers) works just fine with the existing merge implementation. I still think we should switch to `Object.assign` but I decided to present that in a new PR #3124 so we can finish that discussion there.

oliviertassinari commented on src/styles/getMuiTheme.js:
```diff
@@ -240,7 +242,9 @@ export default function getMuiTheme(baseTheme, muiTheme) {
     },
   }, muiTheme);
 
+  const transformers = [autoprefixer, rtl, callOnce].map(t => t(muiTheme)).filter(t => t);
```
What about adding the callOnce only if `process.env.NODE_ENV !== 'production'`?

newoga replied:
That's already happening, remember we change the implementation to do the production check within the transformer transformers/callOnce.js#L6. 

If that method returns undefined, then it is not included in the transfomer chain.  The `filter` method on this line of code is filtering all transformer "creators" that are returning a falsy value (such as when the environment is in production).

newoga replied:
It's essentially the same thing, just a little bit more functional.

oliviertassinari replied:
Ohhh, yes sorry.

newoga replied:
No worries! You're a beast reviewing PRs. Thank you for your endurance. :smile:

oliviertassinari commented on src/styles/transformers/callOnce.js:
```diff
@@ -0,0 +1,15 @@
+const CALLED_ONCE = 'muiPrepared';
+
+export default function callOnce() {
+  if (process.env.NODE_ENV === 'production') return;
```
What about englobing the all code inside the `if` statement.
I'm not sure uglify is smart enough to remove unreachable code?

newoga replied:
Can do!

oliviertassinari commented on package.json:
```diff
@@ -31,6 +31,7 @@
   "homepage": "http://material-ui.com/",
   "dependencies": {
     "inline-style-prefixer": "^0.6.6",
+    "lodash.flowright": "^3.2.1",
```
What about upgrading to `v4`?

newoga replied:
I think 3.2.1 is still the latest version of the `lodash.flowright` package in npm. Not sure exactly what that means though!

oliviertassinari replied:
That's right and weird. Let's keep it like it.

alitaheri replied:
@oliviertassinari lodash only bumps it's tiny packages if they actually suffer a change in a way. that means even if lodash goes to v5 it may still remain like this and then sudenly bump to v6. I think its very clever and intuitive of them to go through such pain to implement that :+1:

oliviertassinari replied:
@alitaheri That's good to know thanks. I'm wondering if babel is doing the same with https://github.com/kittens/lerna.

oliviertassinari commented:
@newoga Yes, let's merge this :tada:

newoga commented:
@oliviertassinari Sounds good! @alitaheri Do you want to give this a last look?

oliviertassinari commented:
@alitaheri or @mbrookes feel free to merge if you are happy with it.

alitaheri commented:
All good and awesome :+1: :+1: JUst fix those 2 things pointed out by @oliviertassinari And do the merge. Really nice work @newoga :+1: :+1:

newoga commented:
Thanks @alitaheri, much appreciated :smile: 

If all is green, this should be good to go.

oliviertassinari commented:
:rocket:!

newoga commented:
I know what a long one! Thanks guys! Big group effort :tada: :balloon:

alitaheri commented:
Really nice work. this was a big step indeed :+1: :+1:
